network/crl.go

```diff
@@ -0,0 +1,380 @@
+/*
+Copyright 2020, Cossack Labs Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/golang/groupcache/lru"
+)
+
+// --tls_crl_from_cert=<use|ignore>
+const (
+	// If certificate contains CRL distribution point(s), use them
+	crlFromCertUseStr = "use"
+	// Ignore CRL distribution points listed in certificate
+	crlFromCertIgnoreStr = "ignore"
+)
+
+const (
+	crlFromCertUse int = iota
+	crlFromCertIgnore
+)
+
+// CRLConfig contains configuration related to certificate validation using CRL
+type CRLConfig struct {
+	uri      string
+	fromCert int // crlFromCert*
+}
+
+// NewCRLConfig creates new CRLConfig
+func NewCRLConfig(uri, fromCert string) (*CRLConfig, error) {
+	_, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	var fromCertVal int
+	switch fromCert {
+	case crlFromCertUseStr:
+		fromCertVal = crlFromCertUse
+	case crlFromCertIgnoreStr:
+		fromCertVal = crlFromCertIgnore
+	default:
+		return nil, errors.New("Invalid `tls_crl_from_cert` value '" + fromCert + "', should be one of 'use', 'ignore'")
+	}
+
+	if uri != "" {
+		// Since this is CRL configuration alone, we don't have access to cache yet;
+		// so let's just download the CRL and forget about it
+		_, err = DefaultCRLClient.Fetch(DefaultCRLClient{}, uri)
+		if err != nil {
+			log.WithError(err).Warnf("CRL: Cannot fetch configured URI '%s'", uri)
+		}
+	}
+
+	return &CRLConfig{uri: uri, fromCert: fromCertVal}, nil
+}
+
+// CRLClient is used to fetch CRL from some URI
+type CRLClient interface {
+	// Fetch fetches CRL from passed URI (can be either http:// or file://)
+	Fetch(uri string) ([]byte, error)
+}
+
+// DefaultCRLClient is a default implementation of CRLClient
+// (as opposed to stub ones used in tests)
+type DefaultCRLClient struct{}
+
+// Fetch fetches CRL from passed URI (can be either http:// or file://)
+func (c DefaultCRLClient) Fetch(uri string) ([]byte, error) {
+	if strings.HasPrefix(uri, "http://") {
+		httpRequest, err := http.NewRequest(http.MethodGet, uri, nil)
+		if err != nil {
+			return nil, err
+		}
+		crlURL, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+		httpRequest.Header.Add("Accept", "application/pkix-crl, application/pem-certificate-chain")
+		httpRequest.Header.Add("host", crlURL.Host)
+		httpClient := &http.Client{}
```

Comment:
will be better if we will initialize client object at DefaultCRLClient initialization and will reuse it.
Plus will be great to set up timeouts for HTTP client to avoid long requests and control it
https://golang.org/pkg/net/http/#Client - We can configure `client.Timeout` parameter by some default value (better to use constant)
Plus we can configure transport object which should be used with all timeouts, and tls.Config if we will need it in a future